src/mbgl/renderer/layers/hillshade_layer_tweaker.cpp

```diff
@@ -0,0 +1,85 @@
+#include <mbgl/renderer/paint_parameters.hpp>
+#include <mbgl/renderer/render_tree.hpp>
+#include <mbgl/style/layers/hillshade_layer_properties.hpp>
+#include <mbgl/util/convert.hpp>
+
+namespace mbgl {
+
+using namespace style;
+
+struct alignas(16) HillshadeDrawableUBO {
+    std::array<float, 4 * 4> matrix;
+    std::array<float, 2> latrange;
+    std::array<float, 2> light;
+};
+static_assert(sizeof(HillshadeDrawableUBO) % 16 == 0);
+
+struct alignas(16) HillshadeEvaluatedPropsUBO {
+    Color highlight;
+    Color shadow;
+    Color accent;
+};
+static_assert(sizeof(HillshadeEvaluatedPropsUBO) % 16 == 0);
+
+static constexpr std::string_view HillshadeDrawableUBOName = "HillshadeDrawableUBO";
+static constexpr std::string_view HillshadeEvaluatedPropsUBOName = "HillshadeEvaluatedPropsUBO";
+
+std::array<float, 2> getLatRange(const UnwrappedTileID& id) {
+    const LatLng latlng0 = LatLng(id);
+    const LatLng latlng1 = LatLng(UnwrappedTileID(id.canonical.z, id.canonical.x, id.canonical.y + 1));
+    return {{static_cast<float>(latlng0.latitude()), static_cast<float>(latlng1.latitude())}};
+}
+
+std::array<float, 2> getLight(const PaintParameters& parameters,
+                              const HillshadePaintProperties::PossiblyEvaluated& evaluated) {
+    float azimuthal = util::deg2radf(evaluated.get<HillshadeIlluminationDirection>());
+    if (evaluated.get<HillshadeIlluminationAnchor>() == HillshadeIlluminationAnchorType::Viewport)
+        azimuthal = azimuthal - static_cast<float>(parameters.state.getBearing());
+    return {{evaluated.get<HillshadeExaggeration>(), azimuthal}};
+}
+
+void HillshadeLayerTweaker::execute(LayerGroupBase& layerGroup,
+                                    const RenderTree& renderTree,
+                                    const PaintParameters& parameters) {
+    const auto& evaluated = static_cast<const HillshadeLayerProperties&>(*evaluatedProperties).evaluated;
+
+    if (layerGroup.empty()) {
+        return;
+    }
+
+#if !defined(NDEBUG)
+    const auto label = layerGroup.getName() + "-update-uniforms";
+    const auto debugGroup = parameters.encoder->createDebugGroup(label.c_str());
+#endif
+
+    if (!evaluatedPropsUniformBuffer) {
+        HillshadeEvaluatedPropsUBO evaluatedPropsUBO = {/* .highlight = */ evaluated.get<HillshadeHighlightColor>(),
+                                                        /* .shadow = */ evaluated.get<HillshadeShadowColor>(),
+                                                        /* .accent = */ evaluated.get<HillshadeAccentColor>()};
+        evaluatedPropsUniformBuffer = parameters.context.createUniformBuffer(&evaluatedPropsUBO,
+                                                                             sizeof(evaluatedPropsUBO));
+    }
+
+    layerGroup.visitDrawables([&](gfx::Drawable& drawable) {
+        drawable.mutableUniformBuffers().addOrReplace(HillshadeEvaluatedPropsUBOName, evaluatedPropsUniformBuffer);
+
+        if (!drawable.getTileID()) {
```

Comment:
I see 10 layer tweakers that do this check but also have some code before this check.

Why `drawable.mutableUniformBuffers().addOrReplace(HillshadeEvaluatedPropsUBOName, evaluatedPropsUniformBuffer);` is done before the check and not after?

Reply:
The tile ID is optional, so should always be checked before it's used.  For most layers, it's expected to always be set.

Attaching the shared UBO to the drawable doesn't depend on the tile ID.  The order is likely different because the layers were built in parallel.  It would be nice to go through and make them look more similar, just for clarity, though.